tool/tsh/db.go

```diff
@@ -62,10 +107,112 @@ func onListDatabases(cf *CLIConf) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	showDatabases(cf.SiteName, databases, activeDatabases, cf.Verbose)
+
+	showDatabases(cf.SiteName, displayUsersForDb, databases, activeDatabases, cf.Verbose)
 	return nil
 }
 
+// getUsersForDb finds allowed and denied users given user's effective roleSet and database in consideration. The result is a user-readable string.
+func getUsersForDb(roleSet services.RoleSet, database types.Database) string {
+	if len(roleSet) == 0 {
+		return ""
+	}
+
+	allowedUsers, deniedUsers := listAllowedDeniedDbUsers(roleSet)
+
+	allUsers := make([]string, 0)
+	allUsers = append(allUsers, allowedUsers...)
+	allUsers = append(allUsers, deniedUsers...)
+
+	usersOk := make([]string, 0)
+	usersBad := make([]string, 0)
+
+	wildcardOk := false
+
+	for _, user := range allUsers {
+		err := roleSet.CheckAccess(database,
+			services.AccessMFAParams{Verified: true},
+			&services.DatabaseUserMatcher{User: user})
+
+		if err == nil {
+			usersOk = append(usersOk, user)
+			if user == types.Wildcard {
+				wildcardOk = true
+			}
+		} else {
+			if user != types.Wildcard {
+				usersBad = append(usersBad, user)
+			}
+		}
+	}
+
+	if len(usersOk) == 0 {
+		return "(none allowed)"
+	}
+
+	if len(usersBad) == 0 || (wildcardOk == false) {
+		return fmt.Sprintf("%v", usersOk)
+	}
+
+	return fmt.Sprintf("%v, except: %v", usersOk, usersBad)
+}
+
+// splitAllowDeny is a helper function for splitting the users allow/deny map (true/false) into separate, sorted slices.
+func splitAllowDeny(inputMap map[string]bool) ([]string, []string) {
+	allows := make([]string, 0)
+	denies := make([]string, 0)
+
+	for entity, allow := range inputMap {
+		if allow {
+			allows = append(allows, entity)
+		} else {
+			denies = append(denies, entity)
+		}
+	}
+
+	sort.Strings(allows)
+	sort.Strings(denies)
+
+	return allows, denies
+}
+
+// listAllowedDeniedDbUsers works on a given role set to return a minimal description of allowed set of usernames.
+// It is biased towards *allowed* usernames; It is meant to describe what the user can do, rather than cannot do.
+// For that reason if the user isn't allowed to pick *any* entities, the output will be empty.
+func listAllowedDeniedDbUsers(set services.RoleSet) ([]string, []string) {
```

Comment:
Named return arguments are helpful in cases like this:
```suggestion
func listAllowedDeniedDbUsers(set services.RoleSet) (allowed []string, denied []string) {
```

Reply:
Also, I feel like this logic should be a part of the services.RoleSet object.